Alert history

alexandre-allard opened the pull request:
**Component**: images, build, salt

**Context**: 

**Summary**:

**Acceptance criteria**: 

---

Closes: #3180

alexandre-allard committed:
images: Add metalk8s-alert-logger container image

This container image is used for alert history.

Refs: #3180

alexandre-allard committed:
images: Add metalk8s-alert-logger

This is a simple HTTP server, listening
on a port (default to 19094), waiting for
HTTP post request from alertmanager.
It then logs the content of these requests
to stdout.

Refs: #3180

alexandre-allard committed:
build: Add build of metalk8s-alert-logger container image

Refs: #3180

alexandre-allard committed:
salt: Add metalk8s-alert-logger deployment

Refs: #3180

alexandre-allard committed:
salt: Improve fluent-bit parser

Remove the date from the logs once parsed by
fluent-bit as we do not need it anymore and
it allows to improve the logs readability.
Plus, it is also need by the alert history
feature, this way we only end up with a JSON
formatted alert in logs which makes it easier
to parse/use by other components.

Refs: #3180

alexandre-allard committed:
salt: Update alertmanager default configuration

We now send all the alerts to the alert-logger
receiver.

Refs: #3180

alexandre-allard committed:
salt: Expose Loki API to metalk8s-ui

We also need to enable CORS to allow Cross Origin
requests coming from the web UI.

Refs: #3180

alexandre-allard committed:
tests: Add a scenario for alert history feature

This scenario checks that we can retrieve the
cluster alerts from the Loki API.

Refs: #3180

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/alert-history` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `feature/alert-history`
* Rebase `feature/alert-history` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

alexandre-allard-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/alert-history` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `feature/alert-history`
* Rebase `feature/alert-history` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/alert-history` into
`development/2.8`
* [w/2.9/feature/alert-history](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/feature/alert-history) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

alexandre-allard committed:
Add CHANGELOG entry for alert history feature

Refs: #3180

bert-e commented:
# History mismatch



Merge commit #f6cf8807ddc2744994944db8283b45725c082587 on the integration branch
`w/2.9/feature/alert-history` is merging a branch which is neither the current
branch `feature/alert-history` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `feature/alert-history` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/alert-history` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `feature/alert-history`
* Rebase `feature/alert-history` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

alexandre-allard committed:
docs: Move architecture diagrams into sub directory

Move all `.uml` to a `diagrams` directory, so the
directory only contains architecture document
which is cleaner.

alexandre-allard committed:
docs: Alert history design document

This document describes how the alert history
feature works and explain why.

Refs: #3180

alexandre-allard committed:
docs: Add alert-history to architecture doc index

Also sort the document by ascii order

gdemonet commented on docs/developer/architecture/alert-history.rst:
```diff
@@ -0,0 +1,285 @@
+We use Loki as the storage backend for alert history because it provides
+several advantages.
+
+First, it allows to easily store the alerts by simply logging them on the
+webhook container output, letting Fluent-bit forward the alerts to it.
+
+Loki uses a NoSQL database, which is better to store JSON documents than
+an SQL one, allowing us to not have to create and maintain a database schema
+for the alerts.
+
+Loki also provides an API allowing us to expose and query these alerts using
+the LogQL language.
+
+Plus, since Loki is already part of the cluster, it saves us from having
+to install, manage and expose a new database.
+
+Using Loki, we also directly benefit from its retention and purge mechanisms,
+making the alerts history retention time automatically aligned with all
+other logs (14 days by default).
```
:trophy:

gdemonet commented on docs/developer/architecture/alert-history.rst:
```diff
@@ -0,0 +1,285 @@
+   Another issue is, since we share the retention configuration with the other
+   logs, it is hard to ensure we will keep enough alert history.
+
+   As for now, there is no retention based on labels, streams, tenant or
+   whatever (on-going discussion `GH Loki #162`_).
```
We need to follow this closely, and maybe contribute. Having tenant-based retention could be really nice for separating different kinds of log streams (not only alerts, but also e.g. Salt events, audit logs, ...).

gdemonet commented on docs/developer/architecture/alert-history.rst:
```diff
@@ -0,0 +1,285 @@
+only the JSON part of the log must be stored to make it easier to use by
+automatic tools.
+
+Expose Loki API
+~~~~~~~~~~~~~~~
+
+The Loki API must be reachable via the web UI, therefore it must be exposed
+through an ingress as it is already done for Prometheus or Alertmanager APIs.
```
We should also start thinking about hiding these behind an OIDC proxy or something, cc @thomasdanan

alexandre-allard replied:
Agree, there is already an issue opened for this (#2667) concerning Prometheus and Alertmanager and I've added a comment to not forget about Loki.

gdemonet commented on docs/developer/architecture/alert-history.rst:
```diff
@@ -0,0 +1,285 @@
+A dedicated Grafana dashboard with the alerts and metrics related to them will
+be added.
+
+.. note::
```
```suggestion
.. todo::
```

gdemonet commented on images/metalk8s-alert-logger/main.go:
```diff
@@ -0,0 +1,45 @@
+
+	for _, alert := range alerts.Alerts {
+		encoded_alert, _ := json.Marshal(alert)
+	        log.Println(string(encoded_alert))
```
Some tab was replaced with spaces apparently:

```suggestion
		log.Println(string(encoded_alert))
```

Hints that we should add this file to `doit lint:go` :wink:

gdemonet commented on salt/metalk8s/addons/alert-logger/deployed/deployment.sls:
```diff
@@ -0,0 +1,53 @@
+  labels:
+    app: metalk8s-alert-logger
+    heritage: metalk8s
+    app.kubernetes.io/name: metalk8s-alert-logger
+    app.kubernetes.io/part-of: metalk8s
+    app.kubernetes.io/managed-by: salt
```
Those labels should also be added to the PodTemplate IMO (maybe not in the selector though)

alexandre-allard replied:
Totally, I forgot to do it. :)

gdemonet commented on salt/metalk8s/addons/alert-logger/deployed/deployment.sls:
```diff
@@ -0,0 +1,53 @@
+    app.kubernetes.io/name: metalk8s-alert-logger
+    app.kubernetes.io/part-of: metalk8s
+    app.kubernetes.io/managed-by: salt
+spec:
```
I see no `spec.replicas`, so we will only ever get one of those loggers (except when the Deployment is doing some rolling update)? Would it make sense to make it configurable? IIRC, you said in the design that since AM will retry sending events to the webhook, I guess it's fine to temporarily lose an instance of this Pod, I just wanted to confirm.

alexandre-allard replied:
Yes, there is only one instance.
It was done on purpose, I don't see the point of having multiple instances of this Pod (I may be wrong).
We can still make it configurable if we want, it should not be a problem.

gdemonet commented on salt/metalk8s/addons/alert-logger/deployed/deployment.sls:
```diff
@@ -0,0 +1,53 @@
+spec:
+  selector:
+    matchLabels:
+      app: metalk8s-alert-logger
```
This `app` label is "deprecated", we should rely on the `app.kubernetes.io/name` equivalent instead. Same in the Service selector.

gdemonet commented on salt/metalk8s/addons/logging/fluent-bit/deployed/configmap.sls:
```diff
@@ -109,7 +113,9 @@ Create fluent-bit ConfigMap:
             [PARSER]
-                Name        docker
-                Format      json
+                Name        container
+                Format      regex
+                Regex       ^(?<time>[^ ]+) (?<stream>stdout|stderr) (?<logtag>[^ ]+) (?<message>.+)$
```
I think this parser should always be used for `containerd`-formatted logs, no? And as such, likely this commit should be backported to the first version where Loki was introduced (2.6, I guess?)

alexandre-allard replied:
Yes, I agree, we should probably backport it.
Will do it once this PR is merged.

alexandre-allard replied:
PR opened here #3276

gdemonet commented on tests/post/steps/test_logging.py:
```diff
@@ -194,6 +194,28 @@ def _check_log_line_exists():
+            result_data = response[0]["data"]["result"][0]["values"]
+        except (IndexError, KeyError):
+            result_data = []
+        assert result_data, "No cluster alerts found in Loki"
```
Don't we want to add an extra check to verify we can see the Watchdog alert there, as suggested in the design? This way, we can also ensure the format of such logs is stable over time.

alexandre-allard replied:
Oh, yes we could, I didn't make it specific to Watchdog just in case it would change at some point, to avoid having to change the test and because the purpose was only to ensure the alert history is working fine.
But, you're right, this way we could also ensure we log valid JSON objects, with specific fields, etc (it would avoid silently breaking the UI in case the object sent by Alertmanager changed due to an upgrade of it).
I will change it.

gdemonet reviewed (changes requested):
Mostly peripheral comments, some things should change in this PR though:

- add the logger Go code to the linter
- adjust the K8s manifests to use standard labels
- adjust the test to check alert format (using the Watchdog "example")

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

alexandre-allard committed:
build: Fix format:go target

We were passing a CmdAction object to target
actions, so doit was not doing anything.
We now pass the execute method of the object.

alexandre-allard committed:
build: Add go linting on every go file

alexandre-allard committed:
lint: Apply go format on all go files

alexandre-allard committed:
docs: Use ubuntu:20.04 to build doc

We need a more recent version of plantuml
to be able to build alert-history diagram
and this version is not shipped in ubuntu:18.04

alexandre-allard committed:
docs: Pre-install missing fonts in entrypoint

There is an issue with mktexpk not being able
to create a directory, so wo create these fonts
prior to launch the build.
Also fix an issue by always changing ownership
of doc build artifacts to avoid having root
owned files when doc build fails.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/alert-history` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `feature/alert-history`
* Rebase `feature/alert-history` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gdemonet commented on images/metalk8s-alert-logger/main.go:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	for _, alert := range alerts.Alerts {
+		encoded_alert, _ := json.Marshal(alert)
```
Maybe add some error handling here?

alexandre-allard replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet commented on images/metalk8s-alert-logger/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+FROM golang:1.16.0-alpine AS builder
```
Go and Alpine versions should likely be exposed as build arguments, don't you think?

alexandre-allard replied:
Hmm, yes, why not, I'm not sure if it will useful, but it does not cost anything to do it.
I'll change it.

alexandre-allard replied:
Done

gdemonet reviewed (dismissed):
Some relatively minor comments, feel free to put them in a debt ticket.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

alexandre-allard commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

gdemonet reviewed (approved):
Don't forget about https://github.com/scality/metalk8s/pull/3191/files#r601234682 :wink:

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
